File: algorithms/Validate Binary Search Tree/Validate Binary Search Tree.cpp
// We know inorder traversal of BST is always sorted, so we are just finding inorder traversal and check whether it is in sorted manner or not, but only using const space using prev pointer.
class Solution {
public:
    TreeNode* prev;
    Solution(){
        prev = NULL;
    }
    bool isValidBST(TreeNode* root) {
        if (root == NULL)
            return true;
        bool a = isValidBST(root->left);
        if (!a)
            return false;
        if (prev != NULL)
        {
            if (prev->val >= root->val)
                return false;
        }
        prev = root;
        return isValidBST(root->right);
    }
};

File: algorithms/Validate Binary Search Tree/Validate Binary Search Tree.java
class Solution {
    public boolean isValidBST(TreeNode root) {
        return dfs(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }
    
    public boolean dfs(TreeNode root, int min, int max) {
        if (root.val < min || root.val > max || (root.val == Integer.MIN_VALUE && root.left != null) ||                    (root.val == Integer.MAX_VALUE && root.right != null)) return false;
        boolean leftRight = true;
        if (root.left == null && root.right == null) return true;
        if (root.left != null) {
            leftRight = dfs(root.left, min, root.val - 1);
        }
        if (root.right != null) {
            leftRight = dfs(root.right, root.val + 1, max) && leftRight;
        }
        return leftRight;
    }
}
